kotlinx-coroutines-core/common/src/sync/ThroughputLimiter.kt

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package kotlinx.coroutines.sync
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.sync.ThroughputCounterEventType.*
+import kotlinx.coroutines.time.*
+import kotlin.time.*
+
+@SinceKotlin("1.5")
+@OptIn(ExperimentalTime::class)
+public interface ThroughputLimiter {
+    /**
+     * Acquires a single permit, blocking until the request can be granted. Tells the amount of time slept, if any.
+     */
+    public suspend fun acquire(): Long
+
+    /**
+     * Acquires the given number of permits, blocking until the request can be granted.
+     */
+    public suspend fun acquire(permits: Int): Long
+
+    /**
+     * Acquires a permit if it can be acquired immediately without delay.
+     */
+    public suspend fun tryAcquire(): Boolean
+
+    /**
+     * Acquires permits if it can be acquired immediately without delay.
+     */
+    public suspend fun tryAcquire(permits: Int): Boolean
+
+    /**
+     * Acquires the given number of permits if it can be obtained without exceeding the specified timeout, or returns false immediately (without waiting) if the permits would not have been granted before the timeout expired.
+     */
+    public suspend fun tryAcquire(permits: Int, timeout: Duration): Boolean
+
+    /**
+     * Acquires a permit if it can be obtained without exceeding the specified timeout, or returns false immediately (without waiting) if the permit would not have been granted before the timeout expired.
+     */
+    public suspend fun tryAcquire(timeout: Duration): Boolean
+
+    /**
+     * Get counts of permit resolutions, granted, denied and such.
+     * @see ThroughputCounterEventType
+     */
+    public fun stats(): Map<ThroughputCounterEventType, Long>
+    public fun resetStats()
+}
+
+/**
+ * Limit throughput of events, per interval, to be at most equal to the argument eventsPerInterval.
+ * When the limit is passed, calls are suspended until the calculated point in time when it's
+ * okay to pass the rate limiter.
+ */
+@SinceKotlin("1.5")
+@OptIn(ExperimentalTime::class)
+public interface IntervalLimiter : ThroughputLimiter
```

Comment:
Clean typing.
Since I am not leaking the Impl classes - they are internal - I want the returned types to be distinct from one another, and match the creation functions.
I like declaring types when using functions to create an object, and I think it would look poorly like this
```
val rateLimiter: ThroughputLimiter = rateLimiter()
val intervalLimiter: ThroughputLimiter = intervalLimiter()
```
compared to what I made:
```
val rateLimiter: RateLimiter = rateLimiter()
val intervalLimiter: IntervalLimiter = intervalLimiter()
```